### PR TITLE
langgraph: add nested interrupts/checkpoints

### PR DESCRIPTION
**Pull request by nfcampos**

### CODE REVIEW

File: libs/langgraph/langgraph/pregel/__init__.py [hunk 1]
```diff
@@ -878,6 +890,12 @@ def stream(
                     },
                 }
 
+            # if we're in a subgraph and have a saved checkpoint,
```
Comment (nfcampos) [inline libs/langgraph/langgraph/pregel/__init__.py:893]: this doesn't seem right, means subgraphs never accept new input after their first run?

Comment (vbarda) [reply libs/langgraph/langgraph/pregel/__init__.py:893]: great point, let me fix

File: libs/langgraph/langgraph/pregel/__init__.py [hunk 2]
```diff
@@ -984,12 +1019,7 @@ def put_checkpoint(metadata: CheckpointMetadata) -> Iterator[Any]:
                     # increment start to 0
                     start += 1
                 else:
-                    # no input is taken as signal to proceed past previous interrupt
-                    checkpoint = copy_checkpoint(checkpoint)
-                    for k in channels:
-                        if k in checkpoint["channel_versions"]:
-                            version = checkpoint["channel_versions"][k]
-                            checkpoint["versions_seen"][INTERRUPT][k] = version
+                    raise ValueError("Received no input")
```
Comment (vbarda) [inline libs/langgraph/langgraph/pregel/__init__.py:1022]: @nfcampos technically we shouldn't even get here, not sure what's the cleanest thing to do here

File: libs/langgraph/langgraph/pregel/__init__.py [hunk 3]
```diff
@@ -1176,6 +1216,9 @@ def put_checkpoint(metadata: CheckpointMetadata) -> Iterator[Any]:
 
                 # set final channel values as run output
                 run_manager.on_chain_end(read_channels(channels, output_keys))
+        except GraphInterrupt as e:
+            run_manager.on_chain_end(e)
```
Comment (vbarda) [inline libs/langgraph/langgraph/pregel/__init__.py:1220]: @nfcampos i don't think this is actually working, i still see error getting logged in langsmith (even though we are catching the interrupt correctly here). am i doing anything wrong?

image

Comment (nfcampos) [reply libs/langgraph/langgraph/pregel/__init__.py:1220]: I think passing an error to on_chain_end is wrong?

Comment (vbarda) [reply libs/langgraph/langgraph/pregel/__init__.py:1220]: yea, fair enough -- what should we pass instead?
i'll update that, but the issue is unrelated

Comment (nfcampos) [reply libs/langgraph/langgraph/pregel/__init__.py:1220]: pass `read_channels(channels, output_keys)` as we do in the line above?

**Review by nfcampos — COMMENTED** [review]
Can you update the checkpointer reference implementations in docs too?

File: libs/langgraph/langgraph/pregel/__init__.py (hunk 2 above)
Comment (nfcampos) [reply libs/langgraph/langgraph/pregel/__init__.py:1024]: Let's create a new exception class for this and use run line 1061 too. eg. `EmptyInputError`

File: libs/langgraph/langgraph/pregel/__init__.py [hunk 4]
```diff
@@ -1108,7 +1143,13 @@ def put_checkpoint(metadata: CheckpointMetadata) -> Iterator[Any]:
                             del fut, task
 
                     # panic on failure or timeout
-                    _panic_or_proceed(done, inflight, step)
+                    # NOTE: for subgraphs we'll raise GraphInterrupt exception on interrupt
+                    exceptions_to_handle = () if is_subgraph else (GraphInterrupt,)
+                    try:
```
Comment (nfcampos) [inline libs/langgraph/langgraph/pregel/__init__.py:1150]: can we do an if statement instead? ie don't use a try statement without exceptions to catch

File: libs/langgraph/langgraph/pregel/__init__.py [hunk 5]
```diff
@@ -945,8 +964,27 @@ def put_checkpoint(metadata: CheckpointMetadata) -> Iterator[Any]:
                             ),
                         )
 
+                if is_subgraph and not checkpoint["channel_values"]:
```
[the remark this answers is not in the document]
Comment (vbarda) [reply libs/langgraph/langgraph/pregel/__init__.py:969]: this is needed to handle the case of parent graph interrupt before subgraph node with interrupt: when we pass `input=None` to resume _parent_ graph, subgraph hasn't run anything yet at that point. so if we set `is_resuming` for subgraph to True at that point, we would try to resume with an empty subgraph checkpoint and break. so we do this check to actually pass inputs to the subgraph in that scenario

let me know if there's a better way to do it!

File: libs/langgraph/langgraph/pregel/__init__.py (hunk 3 above)
Comment (vbarda) [reply libs/langgraph/langgraph/pregel/__init__.py:1222]: but we don't have access to `ChannelsManager` at that point, should we add it there?

File: libs/langgraph/langgraph/pregel/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nfcampos) [reply libs/langgraph/langgraph/pregel/__init__.py:969]: I see, okay, I'll reorganize the code a bit, but makes sense

File: libs/langgraph/langgraph/pregel/__init__.py [hunk 6]
```diff
@@ -1616,12 +1615,14 @@ def put_checkpoint(metadata: CheckpointMetadata) -> Iterator[Any]:
...
-                    try:
-                        _panic_or_proceed(done, inflight, step, asyncio.TimeoutError)
-                    except exceptions_to_handle:
-                        break
+                    if is_subgraph:
+                        _panic_or_proceed(done, inflight, step)
+                    else:
+                        # NOTE: for subgraphs we'll raise GraphInterrupt exception on interrupt
+                        try:
+                            _panic_or_proceed(done, inflight, step)
```
Comment (nfcampos) [inline libs/langgraph/langgraph/pregel/__init__.py:1623]: missing asyncio.TimeoutError

File: libs/langgraph/langgraph/pregel/__init__.py [hunk 7]
```diff
@@ -1691,8 +1692,10 @@ def put_checkpoint(metadata: CheckpointMetadata) -> Iterator[Any]:
 
                 # set final channel values as run output
                 await run_manager.on_chain_end(read_channels(channels, output_keys))
-        except GraphInterrupt as e:
-            await asyncio.shield(run_manager.on_chain_end(e))
+        except GraphInterrupt:
+            await asyncio.shield(
```
Comment (nfcampos) [inline libs/langgraph/langgraph/pregel/__init__.py:1696]: don't think we need shield here? only when handling cancelled error

File: libs/langgraph/tests/test_pregel.py [hunk 8]
```diff
@@ -7514,7 +7514,18 @@ def side(state: State):
...
+    [
+        MemorySaverAssertImmutable(),
+        SqliteSaver.from_conn_string(":memory:"),
+    ],
+    ids=[
+        "memory",
+        "sqlite",
+    ],
+)
+def test_nested_graph_interrupts(checkpointer: BaseCheckpointSaver) -> None:
```
Comment (nfcampos) [inline libs/langgraph/tests/test_pregel.py:7528]: line 7567 shadows this

File: libs/langgraph/tests/test_pregel.py [hunk 9]
```diff
@@ -7640,7 +7651,18 @@ def outer_2(state: State):
...
+    [
+        MemorySaverAssertImmutable(),
+        SqliteSaver.from_conn_string(":memory:"),
+    ],
+    ids=[
+        "memory",
+        "sqlite",
+    ],
+)
+def test_nested_graph_interrupts_parallel(checkpointer: BaseCheckpointSaver) -> None:
```
Comment (nfcampos) [inline libs/langgraph/tests/test_pregel.py:7665]: line 7705 shadows

File: libs/langgraph/langgraph/pregel/__init__.py (hunk 6 above)
Comment (vbarda) [reply libs/langgraph/langgraph/pregel/__init__.py:1623]: oops, something with went wrong when i merged, thanks

**Comment by nfcampos** [comment]
To fix issue when adding call to put_checkpoint: move all checkpoint updates from prepare_next_tasks to apply_writes

File: libs/langgraph/tests/test_pregel_async.py [hunk 10]
```diff
@@ -6006,6 +6006,381 @@ async def side(state: State):
...
+    [
+        MemorySaverAssertImmutable(),
+        AsyncSqliteSaver.from_conn_string(":memory:"),
+    ],
+    ids=[
+        "memory",
+        "sqlite",
+    ],
+)
+async def test_nested_graph_interrupts(checkpointer: BaseCheckpointSaver) -> None:
```
Comment (nfcampos) [inline libs/langgraph/tests/test_pregel_async.py:6020]: whenever using sqlite saver you need to close it at end of test, see other tests

**Comment by nfcampos** [comment]
Superseded by #1075

**Pull request closed without merging**